### PR TITLE
Console: Add timing information for steps and stages.

### LINKED ISSUE #146: show stage basic information
### What feature do you want to see added?

When user click in pipeline-graph on stage, he will be navigated to pipeline-steps. So far so good, but the user see the message *Select a node to view console output.*

image

This might be true, because you has not selected any steps. But it will be fine, when are here some stage information, like

**Stage 'Install WinCC OA'**
Started 18 min ago
Took 9 min 31 sec
Status : (Success / Failed / Unstable ..)
Link to last one step (in my case I have ~ 100 steps)

optional (nice to have):
On node env.NODE_NAME
count of steps ?
exception - link to step (the last one) which produce a exception (why is the stage not GREEN)

### Upstream changes

_No response_

### PR DESCRIPTION
**Pull request by timbrown5**
This is a mock-up with hardcoded details - to try and iterate onto a nicer solution.

Adds a tooltip which contains some Pipeline details.
Fixes #146 

- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira

### CODE REVIEW

**Comment by timja** [comment]
What's remaining to take this out of draft?

Screenshots in issue looked great

**Comment by timbrown5** [comment]
The step timings need to be fixed - I've looks at blue ocean but can't find
where the step timing is set (mainly step duration).
There is probably some cleanup required too (as I went though various
iterations).

**Comment by mikedld** [comment]
If I were to guess I'd say it's this part in `handleChunkDone`: https://github.com/jenkinsci/blueocean-plugin/blob/c6407e6e25c671eb29fec420a1e2d8f92137a079/blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/PipelineNodeGraphVisitor.java#L222-L248 (the use of `StatusAndTiming` in particular).

**Comment by timbrown5** [comment]
> If I were to guess I'd say it's this part in handleChunkDone: https://github.com/jenkinsci/blueocean-plugin/blob/c6407e6e25c671eb29fec420a1e2d8f92137a079/blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/PipelineNodeGraphVisitor.java#L222-L228 (the use of StatusAndTiming in particular).

I think that is to get timing information for stages? I think the TimingInfo here gets the right values, I was just using them incorrectly. I think it's correct now:

Screenshot 2022-10-17 at 14 30 26

Screenshot 2022-10-17 at 14 30 33

File: src/main/frontend/pipeline-console-view/pipeline-console/main/PipelineConsole.tsx [hunk 1]
```diff
@@ -27,6 +35,70 @@ export interface ConsoleLineProps {
...
+  <React.Fragment>
+    <div className="stage-detail-group">
+      <Typography color="inherit" className="detail-element-header">Stage '{props.stage.name}'</Typography>  
+      <div className="detail-element" key="start-time"><ScheduleIcon className="detail-icon"/>{props.stage.startTimeMillis}</div>
+      <div className="detail-element" key="paused-duration"><HourglassEmptyIcon className="detail-icon"/>{props.stage.pauseDurationMillis}</div>
+      <div className="detail-element" key="duration"><TimerIcon className="detail-icon"/>{props.stage.totalDurationMillis}</div>
+      <div className="detail-element" key="status"><InfoIcon className="detail-icon "/><span className="capitalize">{props.stage.state}</span></div>
+      {
+        props.failedSteps.map((value: StepInfo) => {
+          console.log(`Found failed step ${value}`)
```
Comment (timja) [inline src/main/frontend/pipeline-console-view/pipeline-console/main/PipelineConsole.tsx:55]:
```suggestion
```

File: src/main/frontend/pipeline-console-view/pipeline-console/main/PipelineConsole.tsx [hunk 2]
```diff
@@ -226,6 +298,88 @@ export class PipelineConsole extends React.Component<
...
+      let stage = this.state.stages[i];
+      if ('' + stage.id == this.state.selected) {
+        console.log(`Found stage node with id ${stage.id}`)
+        // User has selected a stage node.
+        focusedStage = stage;
+        let failedSteps = [] as StepInfo[];
+        for (let i = 0; i < this.state.steps.length; i++) {
+          let step = this.state.steps[i];
+          if (step.stageId === this.state.selected) {
+            console.log(`Checking status of step ${step.id} - ${step.state.toLowerCase()}`)
```
Comment (timja) [inline src/main/frontend/pipeline-console-view/pipeline-console/main/PipelineConsole.tsx:313]:
```suggestion
```

File: src/main/frontend/pipeline-console-view/pipeline-console/main/PipelineConsole.tsx [hunk 3]
```diff
@@ -226,6 +298,88 @@ export class PipelineConsole extends React.Component<
...
+            // We seem to get a mix of upper and lower case states, so normalise on lowercase.
+            if (step.state.toLowerCase() === "unstable") {
+              console.log(`Found failed step steps ${step}`)
```
Comment (timja) [inline src/main/frontend/pipeline-console-view/pipeline-console/main/PipelineConsole.tsx:316]:
```suggestion
```

File: src/main/frontend/pipeline-console-view/pipeline-console/main/PipelineConsole.tsx [hunk 4]
```diff
@@ -226,6 +298,88 @@ export class PipelineConsole extends React.Component<
...
+      // Return empty div
+      <div></div>
+    )
+  }
+
+  renderStepDetails() {
+    for (let i = 0; i < this.state.steps.length; i++) {
+      let step = this.state.steps[i];
+      if ('' + step.id == this.state.selected) {
+        console.log(`Found steps node with id ${step.id}`)
```
Comment (timja) [inline src/main/frontend/pipeline-console-view/pipeline-console/main/PipelineConsole.tsx:338]:
```suggestion
```

File: src/main/java/io/jenkins/plugins/pipelinegraphview/utils/PipelineStepVisitor.java [hunk 5]
```diff
@@ -263,7 +263,9 @@ protected void handleChunkDone(@NonNull MemoryFlowChunk chunk) {
             new FlowNodeWrapper(
                 step,
                 new NodeRunStatus(BlueRun.BlueRunResult.UNKNOWN, BlueRun.BlueRunState.QUEUED),
-                new TimingInfo(),
+                // This is a fake AtomNode to help track the cause of a blockage, timeing info is
```
Comment (timja) [inline src/main/java/io/jenkins/plugins/pipelinegraphview/utils/PipelineStepVisitor.java:266]:
```suggestion
                // This is a fake AtomNode to help track the cause of a blockage, timing info is
```

**Comment by timbrown5** [comment]
No, will remove them 👍

**Pull request merged**